File: problems/cpp/permutations.cc
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums)
    {
        vector<vector<int>> result;
        permute_recursive(0, &nums, &result);
        return result;
    }
private:
    void permute_recursive(int depth, vector<int>* arr, vector<vector<int>>* result)
    {
        if (depth == arr->size() - 1) {
            result->emplace_back(*arr);
            return;
        }
        for (int i = depth, n = arr->size(); i < n; ++i) {
            swap((*arr)[depth], (*arr)[i]);
            permute_recursive(depth + 1, arr, result);
            swap((*arr)[depth], (*arr)[i]);
        }
    }
};

File: code/python/permutations.py
# 03/16/2018
# Time  : O(2**n)
# Space : O(n)

class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        def permute_recursive(index):
            if index == len(nums):
                result.append(copy.copy(nums))
                return
            for i in range(index, len(nums)):
                nums[i], nums[index] = nums[index], nums[i]
                permute_recursive(index + 1)
                nums[i], nums[index] = nums[index], nums[i]
        
        permute_recursive(0)
        return result
